README.md
cpp-ORM
=========

Current build status : 
[![Build Status](https://travis-ci.org/Krozark/cpp-ORMg.png?branch=master)](https://travis-ci.org/Krozark/cpp-ORM)

An ORM project.

You can simply create persistent objects using databases.

The object representation:
    Each object have to be in a separate table with a pk column named 'id' as auto increment.
    each attr is a column in this table.

Table ares created automaticly.


Tested on
----------

* Windows x64 (sqlite3 only) + Mingw 4.9 [v 0.4+]
* Ubuntu + Gcc 4.8+ [v 0.1+]
* Ubuntu + clang 3.3+ [v 0.1+]



functions
---------

* Persistent Object
 * print as json
 * save / update / load
 * polymorphic object [0.4.2]
* Foreign key
 * save / update / load
* ManyToMany
 * save / update / load / add / remove
* filters (and, or, not)
 * WHERE statement
 * and, or, not 
 * order by
 * limit
* caching
* debug output (set ORM_DEBUG to ORM_DEBUG_XXX in debug.hpp)
* multi connection for a single object (select/save) [v 0.3]
* Table creation, clearing and drop

Data bases supported
=====================

* MySql
* Sqlite3


Requirement
===========

* mysql-connector-c (if Mysql support is needed). Can be find here : https://dev.mysql.com/downloads/connector/c/
* ~~lib Sqlite3~~ [included since v 0.4]
* pthread (for linux only) [v 0.4+]
* doxygen (for user doc only)


Installation
===========

You just have to use cmake:

    mkdir build
    cd build
    cmake ..
    make
    make install

You can customise the backends supported using

* ORM_BUILD_SUPPORT_SQLITE3
* ORM_BUILD_SUPPORT_MYSQL

Exemple
=======

You can see complet exemple in main.cpp

To build the exemple (the database is include)

    set the BUILD_EXAMPLES to True with cmake


```C++
 
    //#include <ORM/backends/MySql.hpp>
    //create your default database
    orm::MySqlDB def("root","root","test");

    //#include <ORM/backends/Sqlite3.hpp>
    //orm::Sqlite3DB def("./datas/test.db"); //if you want to use sqlite3

    //make it as default
    orm::DB& orm::DB::Default = def;

    
    #include <ORM/fields.hpp>
    #include <ORM/models/SqlObject.hpp>
    //create you class
    class Perso : public orm::SqlObject<Perso>
    {
        public:
            Perso();
            orm::CharField<255> name;
            orm::IntegerField pv;
            orm::IntegerField lvl;
    
            MAKE_STATIC_COLUM(name,pv,lvl)
    };
    REGISTER_AND_CONSTRUCT(Perso,"perso",name,"nom",pv,"pv",lvl,"lvl")

    int main(int argc,char* argv[])
    {
        //connect to database
        orm::DB::Default.connect();
    
        //get perso with pk=1
        Perso::type_ptr p1 = Perso::get(1);
        //see it
        cout<<*p1<<endl;
        //modify it
        p1->pv +=14;
        //save it
        p1->save();


        //add one
        Perso::type_ptr p2 = Perso::create();// = new Perso;
        p2->name = "test insert";
        p2->lvl = 75;
        p2->save(); //save it
        cout<<*p2<<endl;

        // all
        cout<<"All persos"<<endl;
        Perso::result_type lis= Perso::all();
        for(auto u : lis)
            cout<<*u<<endl;

        list.clear();
        //custom query
        Perso::query()\
            .filter(Q<Perso>(4,"gt",Perso::$lvl)
                and Q<Perso>(42,"gte",Perso::$lvl)
                and Q<Perso>("test","startswith",Perso::$name)
                and not Q<Perso>(4,"lt",Perso::$lvl)
            ).orderBy(Perso::$name)\
            .limit(42)\
            .get(lis); //get take a list T::result_type& or a T& as param
        for(auto u : lis)
            cout<<*u<<endl;


        orm::DB::Default.disconnect();
        return 0;
    }

```

Class
=====

All class are under the orm namespace


orm::SqlObject<T>
-----------------

Base class for persistent T class. It add all static method / attribute to your class.

make some static method:
* T::result_type get(int pk)
* T::result_type all()
* orm::QuerySet\<T\> T::query()  construct a queryset to make custom query

static member:
* DB* default_connection : database where the object is contain
* std::string table : table name on the database

make member fonction:
* bool save(bool recursive=false,DB&=*default_connection)
* bool del(bool recursive=false,DB&=*default_connection)

Exemple: see main.cpp

Fields
------

* AutoField
* BooleanField
* CharField\<int\>
* TextField
* DateTimeField
* AutoDateTimeField
* AutoNowDateTimeField
* DoubleField
* FloatField
* IntegerField


* FK\<T\>
* ManyToMany\<OWNER,T\>

orm::Attr\<T\>
--------------

Make a T persistent in the database.

You can access to the value using:
* T value;


All operators are overloads to make a easy use of it.

You may prefer to use premake fields

orm::ManyToMany\<T,U\>
======================

T have to be the persistent class where the M2M is use, and U the persistent class.
I create a link with the 2 class, and you can access to the U using:

accessors:
* U::result_type .all()


orm::DB
---------

Base classe to deal with your database. Use specialize database (MySqlDB, ...)

Constructor:
DB(const std::string& username,const std::string& pass,const std::string& db,const std::string& serveur,const std::string& port);


You can use some fonction:
* bool connect()
* bool disconnect()
* Query* query()
* Query* query(std::string&)
* Query* query(std::string&&)
* Query* preparedQuery(std::string&)
* Query* preparedQuery(std::string&&)


You can custom the default database:

    orm::MySqlDB def("root","root","test");
    orm::DB& orm::DB::Default = def;

Macros
======

There are some macos to help.

MAKE_STATIC_COLUM(...)
----------------------

To call in your class with all persistent attrs in params. It will construct the static std::string to use (on filter ...).
ORM_MAKE_NAME(colum) turn your colunm name to the on to use by adding ORM_COLUMN_CHAR ($ by default)

exemple:

```C++
    class Spell : public orm::SqlObject<Spell>
    {
        public:
            Spell();
            orm::CharField<255> name;
            orm::IntegerField element;

            MAKE_STATIC_COLUMN(name,element);

    };
```

the colums $name and $element will be created for query.

```C++
    Spell::
    auto results = Spell::query().filter("spell name",orm::op::exact,Spell::$name);
```



REGISTER_AND_CONSTRUCT(klass,colum,[attr,colum]+)
-------------------------------------------------

Make the default constructor of your class, construct all comulm as static strings.
This macro simply call REGISTER_TABLE(klass,colum) and MAKE_CONSTRUCTOR(klass,__VA_ARGS__)


Marcos II
---------

Only to use if you want build a custum constructor



REGISTER_TABLE(klass,colum)
------------------------------

If you use REGISTER_AND_CONSTRUCT, you don't need to use it.

Thiis will create all the static variables needed by the library


MAKE_CONSTRUCTOR(klass[,attr,colum]+)
--------------------------------------

If you use REGISTER_AND_CONSTRUCT, you don't need to use it.

Make the default constructor for the class klass, and register all Att in params. It also call _MAKE_STRING_N to create attrs colum names as static.

     _MAKE_STRING_N(klass,NUM_ARGS(__VA_ARGS__),__VA_ARGS__)
    klass::klass(): _MAKE_ATTRS_N(NUM_ARGS(__VA_ARGS__),__VA_ARGS__)
    {
         _MAKE_REGISTER_ATTRS(NUM_ARGS(__VA_ARGS__),__VA_ARGS__)
    }


 _MAKE_STRING_N(klass,N,...)
----------------------------
If you use REGISTER_AND_CONSTRUCT, you don't need to use it.

N have to be the number of attrs (in ...)

Simply construct all string as:
    std::string klass::_attr = colum;


_MAKE_ATTRS_N(N,...)
--------------------
If you use REGISTER_AND_CONSTRUCT, you don't need to use it.

N have to be lenght of ...

simply add attr(colum) to constuctor for const initialize


_MAKE_REGISTER_ATTRS(N,...)
---------------------------------------
If you use REGISTER_AND_CONSTRUCT, you don't need to use it.

for each att in ...

    this->registerAttr(this->name);


changelog
0.4.2
=====

MySqlDB
-------

* Use C API instead of C++

SqlExtends
----------

* Add this class to support polymorphic object.

0.4.1
=====

Query
------

* Change column prefix from '_' to '$' to avoid double underscore in variable name that can cause undefine behaviours.
* Add alias for query return type Queryset\<T\>::result_type , Sqlobject\<T\>::result_type

SqlObjectBase
-------------

* Add [before/after]_load methods. You can now add some logical befor/after an object is created from the DB 


0.4.0
=====

Fields
------

* Update DateTimeFiled for windows support

Backends
--------

* Remove extarnals dependencies for sqlite3

Cmake
-----

* Add option for each backends such as
 * BUILD_SUPPORT_SQLITE3
 * BUILD_SUPPORT_MYSQL

0.3.5
=====

Fields
-------

* Add DateTimeField
* Add AutoDateTimeField
* Add AutoNowDateTimeField

Add [before/after]_[save/update] methods

You can now add some logical befor/after a BD query is made 

SqlObjectBase
-------------

* Add [before/after]_[save/update] methods. You can now add some logical before/after a BD query is made 


0.3.4:
======

Add M2MRegister
Add M2MFilter


main.cpp
//#include <ORM/backends/MySql.hpp>
//orm::MySqlBdd def("root","toor","test");
//orm::MySqlBdd def("root","root","test");

#include <ORM/backends/Sqlite3.hpp>
orm::Sqlite3Bdd def("./datas/test.db");

orm::Bdd& orm::Bdd::Default = def;

/**
 * \todo ADD transaction (start transaction / commit)
 **/

#include <ORM/fields.hpp>
#include <ORM/fields/ManyToMany.hpp>
#include <ORM/models/SqlObject.hpp>

#include <iostream>


class Spell : public orm::SqlObject<Spell>
{
    public:
        Spell();
        orm::CharField<255> name;
        orm::IntegerField element;

        MAKE_STATIC_COLUMN(name,element);

};
REGISTER_AND_CONSTRUCT(Spell,"spell",name,"name",element,"element")

class Stats : public orm::SqlObject<Stats>
{
    public:
        Stats();
        orm::IntegerField pv;
        orm::IntegerField pi;
        /*orm::IntegerField intelligence;
        orm::IntegerField force;
        orm::IntegerField defence;
        orm::IntegerField vattaque;
        orm::IntegerField esquive;
        orm::IntegerField chance;
        orm::IntegerField charme;
        orm::IntegerField mouvement;*/

        MAKE_STATIC_COLUMN(pv,pi/*,intelligence,force,defence,vattaque,esquive,chance,charme,mouvement*/)
};
REGISTER_AND_CONSTRUCT(Stats,"stats",pv,"pv",pi,"pi"/*,intelligence,"int",force,"force",defence,"def",vattaque,"vatq",esquive,"esq",chance,"chance",charme,"charme",mouvement,"mouvement"*/)

class Perso : public orm::SqlObject<Perso>
{
    public:
        Perso();
        orm::CharField<255> name;
        orm::IntegerField lvl;
        orm::FK<Stats,false> stats;
        orm::FK<Stats,false> stats2;
        orm::FK<Perso,true> maitre;

        orm::ManyToMany<Perso,Spell> spells;

        MAKE_STATIC_COLUMN(name,lvl,stats,stats2,maitre)
};
//REGISTER_AND_CONSTRUCT(Perso,"perso",name,"name",lvl,"lvl",stats,"stats",stats2,"stats_tmp",maitre,"master")
M2M_REGISTER(Perso,spells,Spell,"perso_spell","perso_id","spell_id")
REGISTER(Perso,"perso",name,"name",lvl,"lvl",stats,"stats",stats2,"stats_tmp",maitre,"master")
Perso::Perso() : name(Perso::_name), lvl(Perso::_lvl), stats(Perso::_stats),stats2(Perso::_stats2), maitre(_maitre), spells(*this)
{
    name.registerAttr(*this);
    lvl.registerAttr(*this);
    stats.registerAttr(*this);
    stats2.registerAttr(*this);
    maitre.registerAttr(*this);
}


class TestTypes : public orm::SqlObject<TestTypes>
{
    public:
    TestTypes();

    orm::AutoField autoField;
    orm::BooleanField booleanField;
    orm::CharField<255> charField;
    orm::IntegerField integerField;
    orm::FloatField floatField;

    MAKE_STATIC_COLUMN(\
                       autoField,\
                       booleanField,
                       charField,
                       integerField,\
                       floatField\
                      )
};
REGISTER_AND_CONSTRUCT(TestTypes,"test_types",\
                       autoField,"autoField",\
                       booleanField,"booleanField",\
                       charField,"charField",\
                       integerField,"integerField",\
                       floatField,"floatField"\
                      )

using namespace orm;
using namespace std;

int main(int argc,char* argv[])
{
    orm::Bdd::Default.connect();
    //REGISTER_BDD(Perso,orm::Bdd::Default)
    
    Bdd* con2 = orm::Bdd::Default.clone();
    con2->connect();

    TestTypes::create();
    TestTypes::clear();
    TestTypes::drop();



    { 
        auto& p1 = Perso::get(1,*con2);
        cout<<"Current perso1 "<<*p1<<endl;
        cout<<" add 1 to lvl"<<endl;
        p1->lvl = p1->lvl + 1;
        cout<<"Current perso1 "<<*p1<<endl;
        cout<<"save it"<<endl;
        cout<<"current lvl: "<<p1->lvl<<endl;
        p1->save();

        cout<<"Current perso1 "<<*p1<<endl;
        cout<<" add 2 to stats.pv"<<endl;
        p1->stats->pv += 2;
        cout<<"Current perso1 "<<*p1<<endl;
        p1->save(*Perso::default_connection,true);
        cout<<"Current perso1 "<<*p1<<endl;

        cout<<"delete p1->master->master"<<endl;
        p1->maitre->maitre.del(*Perso::default_connection,true);

        cout<<"Current perso1 "<<*p1<<endl;
    }

   {
       cout<<"All persos"<<endl;
       std::list<Cache<Perso>::type_ptr> lis= Perso::all();
       for(auto u : lis)
           cout<<*u<<endl;
   }

   {
       cout<<"Create Perso"<<endl;
       Perso p2;// = new Perso;

       p2.name = "test insert";
       p2.lvl = 75;
       p2.save();

       cout<<"Change PV to +=20"<<endl;
       p2.stats->pv+= 20;
       cout<<p2<<endl;

       cout<<"save it"<<endl;
       //p2.save();

       cout<<"All persos (current="<<p2<<")"<<endl;
       std::list<std::shared_ptr<Perso> > lis= Perso::all();
       for(auto u : lis)
           cout<<*u<<endl;

    }
   {
       Perso p2;

       std::list<Cache<Perso>::type_ptr> results;
       Perso::query()\
           .filter(4,"gt",Perso::_lvl)\
           .filter(42,"gte",Perso::_lvl)\
           .exclude(4,"lt",Perso::_lvl)\
           .orderBy(Perso::_name)\
           .get(p2);

       std::cout<<p2<<std::endl;

       cout<<"All perso"<<endl;
       Perso::query().get(results);
       for(auto& perso : results)
           cout<<*perso<<endl;

       results.clear();

       cout<<"All perso where lvl < 200"<<endl;
       Perso::query().filter(200,"lt",Perso::_lvl).get(results);
       for(auto& perso : results)
           cout<<*perso<<endl;


   }
   {

       std::list<std::shared_ptr<Perso>> lis;
       Perso::query().filter(string("test"),"startswith",Perso::_name).get(lis);
       for(auto u : lis)
           cout<<*u<<endl;
   }



   { 
       auto& p1 = Perso::get(1);
       const std::list<std::shared_ptr<Spell> >& spells = p1->spells.all();
       for(auto u : spells)
           cout<<*u<<endl;
   }
       

    {
       std::list<std::shared_ptr<Spell>> list;
       Spell::query().get(list);
       for(auto u : list)
           cout<<*u<<endl;
    }

    {
        auto& p1 = Perso::get(1);
        std::cout<<"perso pk=1:\n"<<*p1<<std::endl;

        auto& stats = Stats::get(1);
        p1->stats = stats;

        std::list<std::shared_ptr<Spell>> list;
        p1->spells.query()\
            .filter(2,"exact",Spell::_element)\
            .get(list);

        cout<<"sorts d'element == 2"<<endl;
        for(auto u : list)
            cout<<*u<<endl;

        cout<<"sorts du perso"<<endl;
        for(auto u : p1->spells.all())
            cout<<*u<<endl;

        auto& sort = Spell::get(1);
        cout<<"sort pk = 1.\n"<<*sort<<endl;

        p1->spells.add(sort);
        p1->spells.remove(sort);

    }



    Bdd::Default.disconnect();
    return 0;
};
